adafruit_emc2101/__init__.py

```diff
@@ -261,17 +281,23 @@ def external_temperature(self):
         temp_msb = self._ext_temp_msb
         full_tmp = (temp_msb << 8) | temp_lsb
         full_tmp >>= 5
-        full_tmp *= 0.125
-
-        return full_tmp
+        if full_tmp == emc2101_regs.TEMP_FAULT_OPENCIRCUIT or full_tmp == emc2101_regs.TEMP_FAULT_SHORT:
+            if self._sensorfault_exceptions:
+                raise ValueError("External Sensor fault")
+            else:
+                return int(full_tmp)
+        else:
+            full_tmp *= 0.125
+        return float(full_tmp)
```

Comment:
So this, and the stuff about sensorfault exceptions, is because there are two values returned by the device that are not temperatures but indicate fault conditions, as you will note. I wanted to be able to detect them individually for the user.

One way is obviously to throw an exception, and the other is to return a specific value (as the device does).

In the specific value case, letting them be cast to float potentially leaves them mangled so I wanted to return the original 'int' value.

The reason for the 'float(..)' is a visual note that the value *is* a float (and ditto for int), to make it clearer what is happening.

I regard the current state as a work in progress - and interested in your thoughts on this.

Reply:
This goes back to my other response where any deviation from expectation should be handled by raising exceptions.  There's a chance with this code that a user will expect a numeric result and not check for type.  If we want to be sure to alert them of something gone awry, it should probably be handled via exception(s).  This will also allow us to just return `full_tmp * 0.125`, which would be a `float` anyway, so no type casting is necessary.

Reply:
Now only uses exceptions.

Reply:
I thought to find some data about mpy file size. Compiling with mpy-cross, running on my M1 Macbook, and starting from the current main branch. I made a makefile to do the rebuild, too.

`$ git log -1
commit 084021060b74c0985032e0b8e6310dab41179c4a (HEAD, main)
Author: Kattni Rembor <kattni@adafruit.com>
Date:   Mon Mar 28 15:52:04 2022 -0400

    Update Black to latest.

    Signed-off-by: Kattni Rembor <kattni@adafruit.com>

$ ls -l *
-rw-rw-r-- 1 ruth ruth  8344 May 10 17:56 emc2101_lut.py
-rw-rw-r-- 1 ruth ruth 11831 May 10 17:56 __init__.py
-rw-rw-r-- 1 ruth ruth  1006 May 10 18:01 Makefile

arm:
total 17
-rw-r--r-- 1 ruth ruth 12135 May 10 18:00 emc2101_lut.mpy
-rw-r--r-- 1 ruth ruth 18302 May 10 18:00 __init__.mpy

bytecode:
total 9
-rw-r--r-- 1 ruth ruth 2203 May 10 18:01 emc2101_lut.mpy
-rw-r--r-- 1 ruth ruth 3870 May 10 18:01 __init__.mpy
`

Move to my current HEAD:

`$ git log -1
commit 093eefaff229d1b7452948e7b06fd8c219aa7cec (HEAD, origin/extra_registers, extra_registers)
Author: rivimey <ruth@ivimey.org>
Date:   Mon May 9 22:19:14 2022 +0100

    Change AttributeError to ValueError as requested

$ ls -l *
-rw-rw-r-- 1 ruth ruth 18482 May 10 18:06 emc2101_lut.py
-rw-rw-r-- 1 ruth ruth  3283 May 10 18:06 emc2101_regs.py
-rw-rw-r-- 1 ruth ruth 17218 May 10 18:06 __init__.py
-rw-rw-r-- 1 ruth ruth  1000 May 10 18:07 Makefile

arm:
total 26
-rw-r--r-- 1 ruth ruth 24015 May 10 18:07 emc2101_lut.mpy
-rw-r--r-- 1 ruth ruth  2756 May 10 18:07 emc2101_regs.mpy
-rw-r--r-- 1 ruth ruth 25749 May 10 18:07 __init__.mpy

bytecode:
total 14
-rw-r--r-- 1 ruth ruth 4565 May 10 18:07 emc2101_lut.mpy
-rw-r--r-- 1 ruth ruth 1137 May 10 18:07 emc2101_regs.mpy
-rw-r--r-- 1 ruth ruth 5429 May 10 18:07 __init__.mpy
`

Reply:
Using the makefile:

all: arm bytecode

clean:
	rm -f $(ARMOBJS) $(BYTEOBJS)

ARMOBJS=arm/__init__.mpy arm/emc2101_lut.mpy arm/emc2101_regs.mpy
BYTEOBJS=bytecode/__init__.mpy bytecode/emc2101_lut.mpy bytecode/emc2101_regs.mpy

arm: $(ARMOBJS)
bytecode: $(BYTEOBJS)

MPYCROSS=$(HOME)/fanctrl.py/bin/mpy-cross

arm/%.mpy: %.py
	@mkdir -p arm
	$(MPYCROSS) -O2 -march=armv7m -X emit=native -o $@  $<

bytecode/%.mpy: %.py
	@mkdir -p bytecode
	$(MPYCROSS) -O2 -X emit=bytecode -o $@  $<

Reply:
Commenting out one of the UnaryStruct variables results in the bytecode dropping by 21 bytes.
Commenting out one of the RWBit variables results in the bytecode dropping by 19 bytes.
Commenting out the group of exception classes changes the bytecode size by 385 bytes (excl invocations of them)